spec/GraphQLOverHTTP.md

```diff
@@ -201,8 +201,38 @@ may not be the GraphQL server but instead some intermediary such as API gateways
 
 ## Status Codes
 
-If the `data` entry in the response has any value other than `null` (when the operation has successfully executed
-without error) then the response should use the 200 (OK) status code.
-If the operation failed before or during execution, due to a syntax error, missing information, validation error
-or any other reason, then response should use the 4XX or 5XX status codes.
-It is recommended to use the same error codes as the [reference implementation](https://github.com/graphql/express-graphql).
+If the response has Content-Type GraphQL and contains a non-null `data` entry,
+then it MUST have status code `2xx`, and it SHOULD have status code `200`.
+
+If the response has Content-Type GraphQL and has status code `2xx`,
+the entry `data` entry must be either:
+- equal to `null` (in case of an execution error)
+- not present (in case of a validation error)
+
+The result of executing a GraphQL operation may contain partial data as well as encountered errors.
+Errors that happen during execution of the GraphQL operation typically become part of the result,
+as long as the server is still able to produce a well-formed response.
+
+In case of errors that completely prevent the successful execution of the request,
+the server SHOULD respond with the appropriate status code depending on the concrete error condition.
+
+There are four types of validation errors:
+
+1. Unparseable or invalid request body, for example `NONSENSE` or `{"qeury": "{__typename}"}`
+2. GraphQL Specification document validation errors (valid syntax, no loops in fragments, etc)
+3. Custom document validations that can be performed before execution (e.g. graphql-depth-limit)
+4. Runtime validations performed by resolvers (i.e. during the execution of the GraphQL operation).
+
+The error types 1. and 2. prevent execution of the request entirely and MUST result in
+status code `400` (Bad Request). Types 3. and 4. may still allow partial execution and
+MUST result in status code `200`.
+
+If an unauthenticated client is not allowed to access the schema, the server MUST respond with
+status code `401` (Unauthorized).
+
+If the server failed unexpectedly, it MUST respond with status code `500` (Internal Server Error).
+
+If a client is not authorized to access any of the requested fields, the server MUST respond with
+status code `403` (Forbidden) and not return any `data`.
```

Comment:
I think this sentence should be deleted in its entirety.

```suggestion
```

Reply:
Thinking about this more, I have come around to agree with you on this. Some fields may error out because of missing authentication, while others may also error out because of other kinds of errors. Rubber-stamping that with a `403` is going too far.